tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc

```diff
@@ -464,6 +464,175 @@ class ConvertFakeQuantWithMinMaxVarsOp : public RewritePattern {
+  /// projection vectors correspond to transformations that do not preserve
+  /// parallelism between lines for example.
+  ///
+  static bool isTranslation(ArrayRef<float> transform, int64_t& rowShift,
+                            int64_t& colShift) {
+    // The transforms is 3x3 matrix with the last entry always one; hence, eight
+    // values.
+    assert(transform.size() == 8 && "unexpected project transforms matrix");
+    // Projection part should all be zero.
+    if (transform[6] != 0 || transform[7] != 0) return false;
+
+    // The top left 2x2 (rotation/scale/skew) part has to be an identity matrix.
+    if (transform[0] != 1 && transform[1] != 0 && transform[3] != 0 &&
+        transform[4] != 1)
+      return false;
+
+    // Shifts. Note that x corresponds to columns, y to rows. The translation is
+    // the negation of the coefficient: for e.g. y' = y - 1 corresponds to a
+    // shift of one (along the +ve y-axis).
+    float colShiftF = -transform[2];
+    float rowShiftF = -transform[5];
+    // Check if these values are integers.
+    if (floorf(rowShiftF) != rowShiftF && floorf(colShiftF) != colShiftF)
+      return false;
+    rowShift = rowShiftF;
+    colShift = colShiftF;
+    return true;
+  }
+
+  LogicalResult matchAndRewrite(ImageProjectiveTransformV3 op,
+                                PatternRewriter& rewriter) const override {
+    // Lower the translation to a pad + slice op. We will pad by the translation
+    // amount and then slice the tensor to its original size. For e.g., a shift
+    // of -2 along the x direction would be a "high" padding of two along the
+    // columns followed by a slice starting at index two.
+    auto outputType = op.getType().dyn_cast<RankedTensorType>();
+    auto inputType = op.images().getType().dyn_cast<RankedTensorType>();
+    if (!outputType || !inputType) {
+      return rewriter.notifyMatchFailure(op,
+                                         "tensors have to be ranked to lower");
+    }
+    assert(inputType.getRank() == 4 && "expected 4-d tensors");
+    if (op.fill_mode() != "CONSTANT") {
+      return rewriter.notifyMatchFailure(op,
+                                         "only CONSTANT fill mode supported");
+    }
+
+    // Get the transformation matrix.
+    Attribute transformsAttr;
+    if (!matchPattern(op.transforms(), m_Constant(&transformsAttr))) {
+      return failure();
+    }
+    auto fpTransformsAttr = transformsAttr.cast<DenseFPElementsAttr>();
+    // Check if the projective transformation is a translation.
+    // The transform is gauranteed to be a float32 tensor. We only support the
+    // case where these is a single transform for all batches. `transforms` can
+    // be [1x8] or [batches x 8].
+    if (fpTransformsAttr.getValues<float>().size() != 8) {
+      return rewriter.notifyMatchFailure(
+          op, "same projective transformation for the entire batch supported");
```

Comment:
Does it means  here the same "type" of protective transformation?

Reply:
No, it means the *same* transformation -- the comment above already clarifies that. Please see: https://www.tensorflow.org/api_docs/python/tf/raw_ops/ImageProjectiveTransformV3

Reply:
Mhh ok cause  we had a quite long thread recently about the within batch randomizzation policy of transformations in the new preprocessing layers. In the case you are interested https://github.com/keras-team/keras-cv/pull/146#issuecomment-1048092586

Reply:
@paynecl @joker-eph How this is going to be optimized in the case we want to randomize a transformation within the batch? 
Could this going to be optimized/fused later if we use it inside a loop?

/cc @LukeWood